File: backend/python/scripts/verify_greeks.py
"""
Greek-profile parity check by finite differences.

For every computable single-expiry strategy we pull the live payoff/greeks
response from the Node API (net Greek curves + the per-leg type/strike/signedQty
the engine actually used), independently rebuild the Black-Scholes portfolio
value V(S, T, r, sigma) from those legs, and numerically differentiate it:

    delta = dV/dS      gamma = d2V/dS2
    vega  = dV/dsigma / 100     (per 1 vol point)
    theta = -dV/dT / 365        (per calendar day)
    rho   = dV/dr / 100         (per 1 rate point)

The finite-difference Greeks are compared, point-by-point on the interior of the
spot grid, against the engine's analytic Greek arrays. This validates both the
Black-Scholes Greek formulas and the per-leg sign/quantity assembly.

Run (from backend/python):
    .\\venv\\Scripts\\python.exe scripts\\verify_greeks.py
"""

import json
import sys
import urllib.request

import numpy as np
from scipy.stats import norm

try:
    sys.stdout.reconfigure(encoding="utf-8")
except Exception:
    pass

NODE = "http://127.0.0.1:4000"

# Multi-expiry structures value the near leg at expiry (non-smooth intrinsic),
# so a smooth-BS finite difference does not apply — checked separately for shape.
MULTI_EXPIRY = {"calendar-call-spread", "calendar-put-spread", "diagonal-call-spread", "diagonal-put-spread"}

# Absolute tolerances (engine rounds delta/theta/vega/rho to 4dp, gamma to 6dp;
# add finite-difference truncation headroom).
TOL = {"delta": 3e-3, "gamma": 3e-3, "vega": 6e-3, "theta": 6e-3, "rho": 6e-3}

# Finite-difference steps.
H_S = 0.10
H_SIG = 1e-4
H_T = 1e-4
H_R = 1e-5


def get(url):
    with urllib.request.urlopen(url, timeout=20) as r:
        return json.load(r)


def post(url, body):
    data = json.dumps(body).encode()
    req = urllib.request.Request(url, data=data, headers={"Content-Type": "application/json"})
    with urllib.request.urlopen(req, timeout=20) as r:
        return json.load(r)


def bs_price(kind, S, K, T, r, sigma):
    S = np.asarray(S, dtype=float)
    if kind == "stock":
        return S.copy()
    if T <= 0 or sigma <= 0:
        return np.maximum(S - K, 0.0) if kind == "call" else np.maximum(K - S, 0.0)
    sq = sigma * np.sqrt(T)
    d1 = (np.log(np.maximum(S, 1e-12) / K) + (r + 0.5 * sigma**2) * T) / sq
    d2 = d1 - sq
    if kind == "call":
        return S * norm.cdf(d1) - K * np.exp(-r * T) * norm.cdf(d2)
    return K * np.exp(-r * T) * norm.cdf(-d2) - S * norm.cdf(-d1)


def portfolio_value(S, T, r, sigma, legs):
    total = np.zeros_like(np.asarray(S, dtype=float))
    for lg in legs:
        q = float(lg["signedQty"])
        total = total + q * bs_price(lg["type"], S, lg["strike"], T, r, sigma)
    return total


def main():
    strategies = get(f"{NODE}/api/strategies?scope=chapter")
    ids = [s["id"] for s in strategies]

    passed, failed, skipped = [], [], []
    print(f"Finite-difference Greek check for {len(ids)} strategies\n")

    for sid in ids:
        detail = get(f"{NODE}/api/strategies/{sid}")
        params = detail["defaultParams"]
        if not detail.get("hasPayoff"):
            skipped.append(sid)
            print(f"  {sid:32s} SKIP (no payoff)")
            continue
        if sid in MULTI_EXPIRY:
            skipped.append(sid)
            print(f"  {sid:32s} SKIP (multi-expiry near-leg intrinsic)")
            continue

        resp = post(f"{NODE}/api/payoff", {"strategyId": sid, "params": params})
        greeks = resp["greeks"]
        agg = greeks["aggregateProfiles"]
        legs = [
            {"type": lg["type"], "strike": lg["strike"], "signedQty": lg["signedQty"]}
            for lg in greeks["legs"]
        ]

        n = len(resp["payoffs"])
        if "spotMin" in params and "spotMax" in params:
            st = np.linspace(float(params["spotMin"]), float(params["spotMax"]), n)
        else:
            st = np.array(resp["spotPrices"], dtype=float)

        T = float(params.get("T", 0.25))
        r = float(params.get("r", 0.05))
        sigma = float(params.get("sigma", 0.25))

        v0 = portfolio_value(st, T, r, sigma, legs)
        vp = portfolio_value(st + H_S, T, r, sigma, legs)
        vm = portfolio_value(st - H_S, T, r, sigma, legs)
        delta_fd = (vp - vm) / (2 * H_S)
        gamma_fd = (vp - 2 * v0 + vm) / (H_S**2)
        vega_fd = (
            portfolio_value(st, T, r, sigma + H_SIG, legs)
            - portfolio_value(st, T, r, sigma - H_SIG, legs)
        ) / (2 * H_SIG) / 100.0
        theta_fd = -(
            portfolio_value(st, T + H_T, r, sigma, legs)
            - portfolio_value(st, T - H_T, r, sigma, legs)
        ) / (2 * H_T) / 365.0
        rho_fd = (
            portfolio_value(st, T, r + H_R, sigma, legs)
            - portfolio_value(st, T, r - H_R, sigma, legs)
        ) / (2 * H_R) / 100.0

        fd = {"delta": delta_fd, "gamma": gamma_fd, "vega": vega_fd, "theta": theta_fd, "rho": rho_fd}

        # Skip 3 points at each end (one-sided edges hurt the 2nd-difference).
        lo, hi = 3, n - 3
        worst = {}
        ok = True
        for g in ("delta", "gamma", "vega", "theta", "rho"):
            eng = np.array(agg[g], dtype=float)[lo:hi]
            num = fd[g][lo:hi]
            d = float(np.max(np.abs(eng - num)))
            worst[g] = d
            if d > TOL[g]:
                ok = False

        line = "  ".join(f"{g[0]}={worst[g]:.1e}" for g in ("delta", "gamma", "vega", "theta", "rho"))
        if ok:
            passed.append(sid)
            print(f"  {sid:32s} OK   {line}")
        else:
            failed.append(sid)
            bad = ",".join(g for g in TOL if worst[g] > TOL[g])
            print(f"  {sid:32s} FAIL [{bad}]  {line}")

    print(f"\nSummary: {len(passed)} match, {len(failed)} mismatch, {len(skipped)} skipped.")
    if failed:
        print("MISMATCHES:", ", ".join(failed))
        sys.exit(1)
    print("All single-expiry Greek profiles match finite differences.")


if __name__ == "__main__":
    main()

File: backend/node/scripts/verify-greeks.mjs
/**
 * Audit BS unit Greeks + portfolio aggregation for every strategy.
 * Run: node scripts/verify-greeks.mjs
 */
import {
  callGreeks,
  callPriceArray,
  putGreeks,
  putPriceArray,
} from "../src/analytics/greeks.js";
import { computePayoff, STRATEGY_IDS } from "../src/analytics/engine.js";
import { NOTEBOOK_BY_ID } from "../src/data/notebookDefaults.js";

const failures = [];
function check(name, cond, detail = "") {
  if (!cond) failures.push(`${name}${detail ? `: ${detail}` : ""}`);
}

const S = 100;
const K = 100;
const T = 0.25;
const r = 0.05;
const sigma = 0.25;

// --- Unit BS identities (Hull / standard European, q=0) ---
const c = callGreeks(S, K, T, r, sigma);
const p = putGreeks(S, K, T, r, sigma);

check("call delta in (0,1)", c.delta > 0.3 && c.delta < 0.7, String(c.delta));
check("put delta in (-1,0)", p.delta > -0.7 && p.delta < -0.3, String(p.delta));
check("Δc − Δp ≈ 1", Math.abs(c.delta - p.delta - 1) < 1e-6, String(c.delta - p.delta));
check("Γc = Γp", Math.abs(c.gamma - p.gamma) < 1e-10, String(Math.abs(c.gamma - p.gamma)));
check("νc = νp", Math.abs(c.vega - p.vega) < 1e-10);
check("Γ > 0", c.gamma > 0 && p.gamma > 0);
check("ν > 0", c.vega > 0 && p.vega > 0);
check("Θc < 0 ATM", c.theta < 0, String(c.theta));
check("ρc > 0", c.rho > 0);
check("ρp < 0", p.rho < 0);
check(
  "ρc − ρp ≈ K T e^{-rT}/100",
  Math.abs(c.rho - p.rho - (K * T * Math.exp(-r * T)) / 100) < 1e-6
);

// Finite-difference delta / gamma / vega / theta / rho
const hS = 0.05;
const callAt = (s, t = T, rr = r, sig = sigma) => callPriceArray([s], K, t, rr, sig)[0];
const putAt = (s, t = T, rr = r, sig = sigma) => putPriceArray([s], K, t, rr, sig)[0];

const fdCallDelta = (callAt(S + hS) - callAt(S - hS)) / (2 * hS);
const fdPutDelta = (putAt(S + hS) - putAt(S - hS)) / (2 * hS);
check("FD call Δ", Math.abs(fdCallDelta - c.delta) < 2e-4, `${fdCallDelta} vs ${c.delta}`);
check("FD put Δ", Math.abs(fdPutDelta - p.delta) < 2e-4, `${fdPutDelta} vs ${p.delta}`);

const fdCallGamma =
  (callAt(S + hS) - 2 * callAt(S) + callAt(S - hS)) / (hS * hS);
check("FD call Γ", Math.abs(fdCallGamma - c.gamma) < 5e-5, `${fdCallGamma} vs ${c.gamma}`);

const hSig = 1e-4;
const fdCallVega = (callAt(S, T, r, sigma + hSig) - callAt(S, T, r, sigma - hSig)) / (2 * hSig) / 100;
check("FD call ν (/100)", Math.abs(fdCallVega - c.vega) < 2e-4, `${fdCallVega} vs ${c.vega}`);

const hR = 1e-4;
const fdCallRho = (callAt(S, T, r + hR) - callAt(S, T, r - hR)) / (2 * hR) / 100;
check("FD call ρ (/100)", Math.abs(fdCallRho - c.rho) < 2e-4, `${fdCallRho} vs ${c.rho}`);

const day = 1 / 365;
const fdCallTheta = (callAt(S, T - day) - callAt(S, T)) / 1; // per calendar day
check("FD call Θ (/day)", Math.abs(fdCallTheta - c.theta) < 5e-3, `${fdCallTheta} vs ${c.theta}`);

// Aggregate must equal sum of leg signed Greeks; profile[S0] ≈ scalar
function nearestIdx(grid, s0) {
  let idx = 0;
  let best = Math.abs(grid[0] - s0);
  for (let i = 1; i < grid.length; i++) {
    const d = Math.abs(grid[i] - s0);
    if (d < best) {
      best = d;
      idx = i;
    }
  }
  return idx;
}

/** Qualitative sign / shape expectations at ATM-ish defaults */
const EXPECT = {
  "long-call": { delta: "+", gamma: "+", vega: "+", theta: "-" },
  "short-call": { delta: "-", gamma: "-", vega: "-", theta: "+" },
  "long-put": { delta: "-", gamma: "+", vega: "+", theta: "?" }, // deep OTM put theta can flip
  "short-put": { delta: "+", gamma: "-", vega: "-" },
  "long-straddle": { delta: "~0", gamma: "+", vega: "+", theta: "-" },
  "short-straddle": { delta: "~0", gamma: "-", vega: "-", theta: "+" },
  "long-strangle": { gamma: "+", vega: "+" },
  "short-strangle": { gamma: "-", vega: "-" },
  "bull-call-spread": { delta: "+" },
  "bear-put-spread": { delta: "-" },
  "bear-call-spread": { delta: "-" },
  "bull-put-spread": { delta: "+" },
  "long-call-butterfly": { gamma: "-", vega: "-" }, // long body short wings? Wait long BF = long wings short body → short gamma ATM
  "short-call-butterfly": { gamma: "+", vega: "+" },
  "long-iron-condor": { gamma: "-", vega: "-", theta: "+" }, // long wings = debit IC? Check naming
  "short-iron-condor": { gamma: "+", vega: "+", theta: "-" },
  "covered-call": { delta: "+", gamma: "-", vega: "-" },
  "protective-put": { delta: "+", gamma: "+", vega: "+" },
  collar: { delta: "+", gamma: "~0" },
  "long-call-synthetic-straddle": { delta: "~0", gamma: "+", vega: "+" },
  "long-put-synthetic-straddle": { delta: "~0", gamma: "+", vega: "+" },
  "short-call-synthetic-straddle": { delta: "~0", gamma: "-", vega: "-" },
  "short-put-synthetic-straddle": { delta: "~0", gamma: "-", vega: "-" },
  strap: { delta: "+" },
  strip: { delta: "-" },
};

function matchSign(val, rule) {
  if (!rule || rule === "?") return true;
  if (rule === "~0") return Math.abs(val) < 0.25;
  if (rule === "+") return val > 0.01;
  if (rule === "-") return val < -0.01;
  return true;
}

for (const id of STRATEGY_IDS) {
  const params = {
    S0: 100,
    K: 100,
    T: 0.25,
    r: 0.05,
    sigma: 0.25,
    ...(NOTEBOOK_BY_ID[id] ?? {}),
  };

  let out;
  try {
    out = computePayoff(id, params);
  } catch (e) {
    check(`${id} compute`, false, e.message);
    continue;
  }

  const { greeks, spotPrices } = out;
  const agg = greeks.aggregate;
  const legs = greeks.legs;
  const idx = nearestIdx(spotPrices, params.S0);

  // Sum of leg scalars == aggregate
  for (const key of ["delta", "gamma", "theta", "vega", "rho"]) {
    const sum = legs.reduce((a, l) => a + Number(l[key]), 0);
    check(
      `${id} aggregate ${key} = Σ legs`,
      Math.abs(sum - agg[key]) < 1e-3,
      `${sum} vs ${agg[key]}`
    );
    const prof = greeks.aggregateProfiles[key][idx];
    check(
      `${id} profile ${key}@S0 ≈ aggregate`,
      Math.abs(prof - agg[key]) < 1e-3,
      `${prof} vs ${agg[key]}`
    );
  }

  // Leg profile length matches spot grid
  for (let i = 0; i < legs.length; i++) {
    check(
      `${id} leg${i} delta profile length`,
      legs[i].profiles.delta.length === spotPrices.length
    );
  }

  // Short legs must flip unit Greek signs relative to long same-type ATM sanity
  for (const leg of legs) {
    if (leg.type === "stock") {
      check(`${id} stock |Δ|≈|signedQty|`, Math.abs(Math.abs(leg.delta) - Math.abs(leg.signedQty)) < 1e-6);
      continue;
    }
    if (leg.direction === "short" && leg.gamma !== 0) {
      // Short options: gamma contribution negative for calls/puts with T>0
      // (expired short legs have gamma 0)
      check(`${id} short ${leg.label} Γ ≤ 0`, leg.gamma <= 1e-9, String(leg.gamma));
    }
    if (leg.direction === "long" && leg.gamma !== 0) {
      check(`${id} long ${leg.label} Γ ≥ 0`, leg.gamma >= -1e-9, String(leg.gamma));
    }
  }

  const exp = EXPECT[id];
  if (exp) {
    for (const [k, rule] of Object.entries(exp)) {
      check(`${id} expected ${k} ${rule}`, matchSign(agg[k], rule), String(agg[k]));
    }
  }

  // Directional latex uses already-signed Δ (no double σ)
  for (const contrib of out.directional.legContributions) {
    check(
      `${id} contrib latex is signed once`,
      contrib.contributionLatex.includes("signed") &&
        Math.abs(Number(contrib.delta) - Number(String(contrib.contributionLatex).match(/([+-]?\d+\.\d+)/)?.[1])) < 1e-3,
      contrib.contributionLatex
    );
  }
}

console.log(`Checked ${STRATEGY_IDS.length} strategies`);
if (failures.length) {
  console.error(`\nFAIL (${failures.length}):`);
  for (const f of failures.slice(0, 80)) console.error(" -", f);
  if (failures.length > 80) console.error(` ... +${failures.length - 80} more`);
  process.exit(1);
}
console.log("\nALL CHECKS PASSED");
